lib/ret/project.ex

```diff
@@ -0,0 +1,56 @@
+defmodule Ret.Project do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  alias Ret.{Repo, Project, OwnedFile}
+
+  @schema_prefix "ret0"
+  @primary_key {:project_id, :id, autogenerate: true}
+
+  schema "projects" do
+    field(:project_sid, :string)
+    field(:name, :string)
+    belongs_to(:created_by_account, Ret.Account, references: :account_id)
+    belongs_to(:project_owned_file, Ret.OwnedFile, references: :owned_file_id)
+    belongs_to(:thumbnail_owned_file, Ret.OwnedFile, references: :owned_file_id)
+
+    timestamps()
+  end
+
+  def project_by_sid(sid) do
+    Project |> Repo.get_by(project_sid: sid)
+  end
+
+  def to_sid(%Project{} = project), do: project.project_sid
+  def to_url(%Project{} = project), do: "#{RetWeb.Endpoint.url()}/projects/#{project |> to_sid}"
+
+  # Create a Project
+  def changeset(%Project{} = project, account, params \\ %{}) do
+    project
+    |> cast(params, [
+      :name
+    ])
+    |> validate_required([
+      :name
+    ])
+    |> validate_length(:name, min: 4, max: 64)
+    # TODO BP: this is repeated from hub.ex. Maybe refactor the regex out.
+    |> validate_format(:name, ~r/^[A-Za-z0-9-':"!@#$%^&*(),.?~ ]+$/)
+    |> maybe_add_project_sid_to_changeset
+    |> unique_constraint(:project_sid)
+    |> put_assoc(:created_by_account, account)
+  end
+
+  # Update a Project with new project and thumbnail files
```

Comment:
it's generally idiomatic to write minimal-covering `changeset` functions -- for example, if we are going to have an access pattern where we *just* update the files, then this changeset function should not have the generic `params` argument. this helps reduce the API surface area, etc.

Reply:
and you'd name this `changeset_for_uploaded_project` or something

Reply:
(afaict that's not the case here, but just mentioning it)

Reply:
Yeah I also allow updating the name, otherwise I could do that.